File: 2025/python/day04.py
from utils import get_data
from utils import parse_grid
from utils import timeit
from utils import DIRS_8


def count_neighbors(grid, point):
    return sum(1 for direction in DIRS_8 if grid.get(point + direction) is not None)


@timeit
def part_one(grid):
    return sum(1 for point in grid if grid[point] < 4)


@timeit
def part_two(grid):
    total_removed = 0
    prev_removed = 0
    while True:
        for point in [key for key, value in grid.items() if value < 4]:
            total_removed += 1
            del grid[point]
            for neighbor in DIRS_8:
                if grid.get(point + neighbor) is not None:
                    grid[point + neighbor] -= 1

        if prev_removed == total_removed:
            return total_removed

        prev_removed = total_removed


@timeit
def main():
    data = get_data("04")
    grid = parse_grid(data, keep_values="@")
    grid = {key: count_neighbors(grid, key) for key in grid}

    p1_result = part_one(grid)
    print(p1_result)

    p2_result = part_two(grid)
    print(p2_result)

File: 2025/rust/day04.rs
use crate::y2025::utils;
use std::collections::HashMap;

fn count_neighbors(grid: &HashMap<utils::Point, char>, point: utils::Point) -> usize {
    point.neighbors_8()
        .iter()
        .filter(|&neighbor| grid.contains_key(neighbor))
        .count()
}

fn part_one(grid: &HashMap<utils::Point, usize>) -> usize {
    grid.values().filter(|&&v| v < 4).count()
}

fn part_two(mut grid: HashMap<utils::Point, usize>) -> usize {
    let mut total_removed = 0;
    let mut prev_removed = 0;

    loop {
        let points_to_remove: Vec<utils::Point> = grid
            .iter()
            .filter(|(_, &v)| v < 4)
            .map(|(&p, _)| p)
            .collect();

        for point in points_to_remove {
            total_removed += 1;
            grid.remove(&point);

            for neighbor in point.neighbors_8() {
                if let Some(count) = grid.get_mut(&neighbor) {
                    *count = count.saturating_sub(1);
                }
            }
        }

        if prev_removed == total_removed {
            return total_removed;
        }

        prev_removed = total_removed;
    }
}

pub fn main() {
    let data = utils::get_data("04");
    let char_grid = utils::parse_grid(&data, Some("@"));

    // Convert grid values to neighbor counts
    let mut grid: HashMap<utils::Point, usize> = HashMap::new();
    for point in char_grid.keys() {
        let count = count_neighbors(&char_grid, *point);
        grid.insert(*point, count);
    }

    let p1_result = part_one(&grid);
    println!("{}", p1_result);

    let p2_result = part_two(grid);
    println!("{}", p2_result);
}
